File: app/src/main/java/com/gigabytedevelopersinc/app/cometOTP/Utilities/DatabaseHelper.java
package com.gigabytedevelopersinc.app.cometOTP.Utilities;

import android.app.backup.BackupManager;
import android.content.Context;
import android.widget.Toast;

import com.gigabytedevelopersinc.app.cometOTP.Database.Entry;
import com.gigabytedevelopersinc.app.cometOTP.R;

import org.json.JSONArray;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;

import javax.crypto.SecretKey;

public class DatabaseHelper {

    static final Object DatabaseFileLock = new Object();

    public static void wipeDatabase(Context context) {
        File db = new File(context.getFilesDir() + "/" + Constants.FILENAME_DATABASE);
        File dbBackup = new File(context.getFilesDir() + "/" + Constants.FILENAME_DATABASE_BACKUP);
        db.delete();
        dbBackup.delete();
    }

    private static void copyFile(File src, File dst)
            throws IOException {
        try (InputStream in = new FileInputStream(src)) {
            try (OutputStream out = new FileOutputStream(dst)) {
                byte[] buffer = new byte[1024];
                int len;
                while ((len = in.read(buffer)) > 0) {
                    out.write(buffer, 0, len);
                }
            }
        }
    }

    public static boolean backupDatabase(Context context) {
        File original = new File(context.getFilesDir() + "/" + Constants.FILENAME_DATABASE);
        File backup = new File(context.getFilesDir() + "/" + Constants.FILENAME_DATABASE_BACKUP);

        if (original.exists()) {
            try {
                copyFile(original, backup);
            } catch (IOException e) {
                return false;
            }
        }

        return true;
    }

    public static boolean restoreDatabaseBackup(Context context) {
        File original = new File(context.getFilesDir() + "/" + Constants.FILENAME_DATABASE);
        File backup = new File(context.getFilesDir() + "/" + Constants.FILENAME_DATABASE_BACKUP);

        if (backup.exists()) {
            try {
                copyFile(backup, original);
            } catch (IOException e) {
                return false;
            }
        }

        return true;
    }

    /* Database functions */
    public static boolean saveDatabase(Context context, ArrayList<Entry> entries, SecretKey encryptionKey) {
        if (encryptionKey == null) {
            Toast.makeText(context, R.string.toast_encryption_key_empty, Toast.LENGTH_LONG).show();
            return false;
        }

        String jsonString = entriesToString(entries);

        try {
            synchronized (DatabaseHelper.DatabaseFileLock) {
                byte[] data = EncryptionHelper.encrypt(encryptionKey, jsonString.getBytes());

                FileHelper.writeBytesToFile(new File(context.getFilesDir() + "/" + Constants.FILENAME_DATABASE), data);
            }
        } catch (Exception error) {
            error.printStackTrace();
            return false;
        }

        BackupManager backupManager = new BackupManager(context);
        backupManager.dataChanged();

        return true;
    }

    public static ArrayList<Entry> loadDatabase(Context context, SecretKey encryptionKey) {
        ArrayList<Entry> entries = new ArrayList<>();

        if (encryptionKey != null) {
            try {
                synchronized (DatabaseHelper.DatabaseFileLock) {
                    byte[] data = FileHelper.readFileToBytes(new File(context.getFilesDir() + "/" + Constants.FILENAME_DATABASE));
                    data = EncryptionHelper.decrypt(encryptionKey, data);

                    entries = stringToEntries(new String(data));
                }
            } catch (Exception error) {
                error.printStackTrace();
            }
        } else {
            Toast.makeText(context, R.string.toast_encryption_key_empty, Toast.LENGTH_LONG).show();
        }

        return entries;
    }

    /* Conversion functions */

    public static String entriesToString(ArrayList<Entry> entries) {
        JSONArray json = new JSONArray();

        for(Entry e: entries){
            try {
                json.put(e.toJSON());
            } catch (Exception error) {
                error.printStackTrace();
            }
        }

        return json.toString();
    }

    public static ArrayList<Entry> stringToEntries(String data) {
        ArrayList<Entry> entries = new ArrayList<>();

        try {
            JSONArray json = new JSONArray(data);

            for (int i = 0; i < json.length(); i++) {
                Entry entry = new Entry(json.getJSONObject(i));
                entries.add(entry);
            }
        } catch (Exception error) {
            error.printStackTrace();
        }

        return entries;
    }
}

File: app/src/main/java/com/gigabytedevelopersinc/app/cometOTP/Utilities/DatabaseHelper.kt
@file:Suppress("PackageName")
package com.gigabytedevelopersinc.app.cometOTP.Utilities

import android.app.backup.BackupManager
import android.content.Context
import android.widget.Toast
import com.gigabytedevelopersinc.app.cometOTP.Database.Entry
import com.gigabytedevelopersinc.app.cometOTP.R
import org.json.JSONArray
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.nio.charset.Charset
import javax.crypto.SecretKey

object DatabaseHelper {

    @JvmField
    internal val DatabaseFileLock = Any()

    @JvmStatic
    fun wipeDatabase(context: Context) {
        val db = File(context.filesDir.toString() + "/" + Constants.FILENAME_DATABASE)
        val dbBackup = File(context.filesDir.toString() + "/" + Constants.FILENAME_DATABASE_BACKUP)
        db.delete()
        dbBackup.delete()
    }

    /** Whether a database has been saved (it can exist and still be unreadable). */
    @JvmStatic
    fun databaseExists(context: Context): Boolean {
        return File(context.filesDir.toString() + "/" + Constants.FILENAME_DATABASE).exists()
    }

    @Throws(IOException::class)
    private fun copyFile(src: File, dst: File) {
        FileInputStream(src).use { `in` ->
            FileOutputStream(dst).use { out ->
                val buffer = ByteArray(1024)
                var len: Int
                while (`in`.read(buffer).also { len = it } > 0) {
                    out.write(buffer, 0, len)
                }
            }
        }
    }

    @JvmStatic
    fun backupDatabase(context: Context): Boolean {
        val original = File(context.filesDir.toString() + "/" + Constants.FILENAME_DATABASE)
        val backup = File(context.filesDir.toString() + "/" + Constants.FILENAME_DATABASE_BACKUP)

        if (original.exists()) {
            try {
                copyFile(original, backup)
            } catch (e: IOException) {
                return false
            }
        }

        return true
    }

    @JvmStatic
    fun restoreDatabaseBackup(context: Context): Boolean {
        val original = File(context.filesDir.toString() + "/" + Constants.FILENAME_DATABASE)
        val backup = File(context.filesDir.toString() + "/" + Constants.FILENAME_DATABASE_BACKUP)

        if (backup.exists()) {
            try {
                // Replaced in one step like saveDatabase() does, not copied over it in place.
                FileHelper.writeBytesToFileAtomically(original, FileHelper.readFileToBytes(backup))
            } catch (e: IOException) {
                return false
            }
        }

        return true
    }

    /* Database functions */
    @JvmStatic
    fun saveDatabase(context: Context, entries: ArrayList<Entry>, encryptionKey: SecretKey?): Boolean {
        if (encryptionKey == null) {
            Toast.makeText(context, R.string.toast_encryption_key_empty, Toast.LENGTH_LONG).show()
            return false
        }

        val jsonString = entriesToString(entries)

        try {
            synchronized(DatabaseFileLock) {
                val data = EncryptionHelper.encrypt(encryptionKey, jsonString.toByteArray(Charset.defaultCharset()))

                // Written to a temporary file and renamed over the database, so a crash mid-write
                // cannot leave a truncated (and then unreadable) database behind.
                FileHelper.writeBytesToFileAtomically(File(context.filesDir.toString() + "/" + Constants.FILENAME_DATABASE), data)
            }
        } catch (error: Exception) {
            error.printStackTrace()
            return false
        }

        val backupManager = BackupManager(context)
        backupManager.dataChanged()

        return true
    }

    /**
     * Loads the database. Returns null when it exists but cannot be read: no key, the wrong key,
     * or a file that cannot be read, decrypted or parsed. A failed load must never be saved or
     * backed up in place of the database, which is why it is not an empty list. A database that
     * does not exist yet (nothing saved so far) loads as an empty list.
     */
    @JvmStatic
    fun loadDatabase(context: Context, encryptionKey: SecretKey?): ArrayList<Entry>? {
        if (encryptionKey == null) {
            Toast.makeText(context, R.string.toast_encryption_key_empty, Toast.LENGTH_LONG).show()
            return null
        }

        synchronized(DatabaseFileLock) {
            return loadDatabaseFile(File(context.filesDir.toString() + "/" + Constants.FILENAME_DATABASE), encryptionKey)
        }
    }

    /** [loadDatabase] without the Context: null when [file] exists but cannot be read. */
    @JvmStatic
    internal fun loadDatabaseFile(file: File, encryptionKey: SecretKey): ArrayList<Entry>? {
        if (!file.exists())
            return ArrayList()

        return try {
            val data = EncryptionHelper.decrypt(encryptionKey, FileHelper.readFileToBytes(file))
            stringToEntriesOrNull(String(data, Charset.defaultCharset()))
        } catch (error: Exception) {
            error.printStackTrace()
            null
        }
    }

    /* Conversion functions */

    @JvmStatic
    fun entriesToString(entries: ArrayList<Entry>): String {
        val json = JSONArray()

        for (e in entries) {
            try {
                json.put(e.toJSON())
            } catch (error: Exception) {
                error.printStackTrace()
            }
        }

        return json.toString()
    }

    /**
     * Parses a JSON array of entries. An entry that cannot be read is skipped on its own; the
     * entries around it are still returned. (Callers save what they loaded, so anything left out
     * here is gone from the database after the next save.)
     */
    @JvmStatic
    fun stringToEntries(data: String?): ArrayList<Entry> {
        if (data == null)
            return ArrayList()
        return stringToEntriesOrNull(data) ?: ArrayList()
    }

    /** [stringToEntries], but null when [data] is not a JSON array at all. */
    @JvmStatic
    internal fun stringToEntriesOrNull(data: String): ArrayList<Entry>? {
        val entries = ArrayList<Entry>()

        val json: JSONArray
        try {
            json = JSONArray(data)
        } catch (error: Exception) {
            error.printStackTrace()
            return null
        }

        for (i in 0 until json.length()) {
            try {
                entries.add(Entry(json.getJSONObject(i)))
            } catch (error: Exception) {
                error.printStackTrace()
            }
        }

        return entries
    }
}
